warehouse/accounts/utils.py

```diff
@@ -0,0 +1,181 @@
+                f"Record is missing attribute(s): {', '.join(missing_keys)}",
+                reason="format",
+            )
+
+        matcher_code = record["type"]
+
+        matcher = matchers.get(matcher_code)
+        if not matcher:
+            raise InvalidTokenLeakRequest(
+                f"Matcher with code {matcher_code} not found. "
+                f"Available codes are: {', '.join(matchers)}",
+                reason="invalid_matcher",
+            )
+
+        try:
+            extracted_token = matcher.extract(record["token"])
+        except ExtractionFailed:
+            raise InvalidTokenLeakRequest(
+                f"Cannot extract token from recieved match", reason="extraction"
+            )
+
+        return cls(token=extracted_token, public_url=record["url"])
+
+
+class TokenLeakAnalyzer:
+    def __init__(self, request):
+        self._request = request
+        self._metrics = self._request.find_service(IMetricsService, context=None)
+        self._macaroon_service = self._request.find_service(
+            IMacaroonService, context=None
+        )
+
+    def analyze_disclosure(self, disclosure_record, origin):
+
+        self._metrics.increment(f"warehouse.token_leak.{origin}.recieved")
+
+        try:
+            disclosure = TokenLeakDisclosureRequest.from_api_record(
+                record=disclosure_record
+            )
+        except InvalidTokenLeakRequest as exc:
+            # TODO Logging something here would be useful in case we recieve
+            # unexpected pattern type.
+            self._metrics.increment(f"warehouse.token_leak.{origin}.error.{exc.reason}")
+            return
+
+        try:
+            database_macaroon = self._macaroon_service.check_if_macaroon_exists(
+                raw_macaroon=disclosure.token
+            )
+        except InvalidMacaroon:
+            self._metrics.increment(f"warehouse.token_leak.{origin}.error.invalid")
+            return
+
+        self._metrics.increment(f"warehouse.token_leak.{origin}.valid")
+        self._macaroon_service.delete_macaroon(macaroon_id=str(database_macaroon.id))
+
+        send_token_compromised_email_leak(
+            self._request,
+            database_macaroon.user,
+            public_url=disclosure.public_url,
+            origin=origin,
+        )
+
+    def analyze_disclosures(self, disclosure_records, origin):
+        if not isinstance(disclosure_records, list):
+            raise InvalidTokenLeakRequest(
+                "Invalid format: payload is not a list", "format"
+            )
+        for disclosure_record in disclosure_records:
+            try:
+                self.analyze_disclosure(
+                    disclosure_record=disclosure_record, origin=origin
+                )
+            except Exception:
+                # TODO log, but don't stop processing other leaks.
+                # It seems logger.exception() is not used in the codebase. What is
+                # expected ?
```

Comment:
Somewhat unfortunate that we might have multiple disclosures to verify, but verifying any one of them may result in an exception (either intentional or unintentional) that may block the verification of one or more disclosures.

This makes me think it might make sense to make each verification it's own background task (via our Celery workers), so each can fail independently of another instead.

Reply:
Yep, sounds nice! I’ll look at that.

Reply:
So just to make sure I got it right:
- Everything that is done only once would be run by the view
- Everything that is done for every individual disclosure would be done in the celery task

Yes ? (I could also imagine having the view launch a main task and then 200, and the main task would do the validation, and create other tasks for each disclosure)